File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 == None:
            return list2
        if list2 == None:
            return list1
        
        cur = None
        head = None # we only use head to ret result
        if list1.val < list2.val:
            cur = list1
            head = list1
            list1 = list1.next
        else:
            cur = list2
            head = list2
            list2 = list2.next

        while list1 != None and list2 != None:
            if list1.val < list2.val:
                cur.next = list1
                cur = cur.next
                list1 = list1.next
            else:
                cur.next = list2
                cur = cur.next
                list2 = list2.next

        # at the end, point it to the leftovers
        if not list1:
            cur.next = list2
        else:
            cur.next = list1
        
        return head

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null) return false;
        ListNode fast = head, slow = new ListNode(-1, head);

        while(slow != fast && fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }

        if(fast == null || fast.next == null){
            return false;
        }

        return true;
    }
}
